src/lib/storage/lz4/lz4_encoder.hpp

```diff
@@ -49,30 +53,35 @@ class LZ4Encoder : public SegmentEncoder<LZ4Encoder> {
     });
 
     /**
-     * Use the LZ4 high compression API to compress the copied values. As C-library LZ4 needs raw pointers as input
-     * and output. To avoid directly handling raw pointers we use std::vectors as input and output. The input vector
-     * contains the data that needs to be compressed and the output vector is allocated enough memory to contain
-     * the compression result. Via the .data() call we can supply LZ4 with raw pointers to the memory the vectors use.
-     * These are cast to char-pointers since LZ4 expects char pointers.
+     * Pre-compute a zstd dictionary if the input data is split among multiple blocks. This dictionary allows
+     * independent compression of the blocks, while maintaining a good compression ratio.
      */
-    DebugAssert(values.size() * sizeof(T) <= std::numeric_limits<int>::max(),
-                "Input of LZ4 encoder contains too many bytes to fit into a 32-bit signed integer sized vector that is"
-                " used by the LZ4 library.");
-
     const auto input_size = values.size() * sizeof(T);
-    // estimate the (maximum) output size
-    const auto output_size = LZ4_compressBound(static_cast<int>(input_size));
-    auto compressed_data = pmr_vector<char>{alloc};
-    compressed_data.resize(static_cast<size_t>(output_size));
-    const int compression_result = LZ4_compress_HC(reinterpret_cast<char*>(values.data()), compressed_data.data(),
-                                                   static_cast<int>(input_size), output_size, LZ4HC_CLEVEL_MAX);
-    Assert(compression_result > 0, "LZ4 compression failed");
-
-    // shrink the vector to the actual size of the compressed result
-    compressed_data.resize(static_cast<size_t>(compression_result));
-    compressed_data.shrink_to_fit();
-
-    return std::allocate_shared<LZ4Segment<T>>(alloc, std::move(compressed_data), std::move(null_values), input_size);
+    auto dictionary = pmr_vector<char>{};
```

Comment:
I personally would prefer using uint8_t over char, since char is not guaranteed to be 8bits and because the semantic of a char is, well, a character, but what you want here is just a byte, without any semantics. To really make that clear you could even do `using byte = uint8_t;` and use byte all the time.

Reply:
We use char, since the underlying libraries (both lz4 and zstd) expect the dictionary to be a char-pointer. A char is guaranteed to be 1 byte large (which is at least 8 bits large).